[RDY] provider/pythonx: only call system("python") once + other fixes

blueyed opened the pull request:
`s:check_interpreter` will query/return the version also, so that
`s:check_version` can just use that, without calling "python" again.

/cc @Shougo

_Update_ I've added some more commits to it afterwards.

blueyed committed:
provider/pythonx: only call system("python") once

`s:check_interpreter` will query/return the version also, so that
`s:check_version` can just use that, without calling "python" again.

Shougo commented:
I get it.  It seems good.

blueyed commented:
Cool!
Renamed it to `[RFC]`, although it is basically `[RDY]` then.

splinterofchaos commented:
LGTM :+1:

blueyed committed:
provider/pythonx: fix grammar with error

blueyed commented:
Thanks!
I've added a new commit to fix the grammar with the error message, and have made the "return" part more readable.

blueyed committed:
provider/pythonx: add reference to ":help nvim-python" for error

blueyed committed:
provider/pythonx: test python3/2 first, do not test python3.2

python3/python2 will typically point at the default / most recent
interpreter.

blueyed committed:
provider/pythonx: Detect: only return valid versions

The `check` return value of `s:check_version` was not being used, and it
was returned always.

blueyed commented:
Regarding the error messages: they are not being used currently anywhere, are they?
There is `provider#python3#Error` / `provider#python#Error`, but it's not used (and defined after the `finish`).

I think they should not show up via `has("python")`, but somewhere - e.g. when calling `:python`?
Currently you just get `E117: Unknown function: provider#python#Call` then.
Maybe this could look for `provider#foo#Error` then? - which would have to be defined in the autoload file then.

blueyed commented:
@Pyrohh 
Good idea, amended - and rebased onto master.

Shougo commented:
> There is provider#python3#Error / provider#python#Error, but it's not used (and defined after the finish).

It is for error debugging.

> I think they should not show up via has("python"), but somewhere - e.g. when calling :python?
Currently you just get E117: Unknown function: provider#python#Call then.

I think it is better.

blueyed commented:
@Shougo 
>> provider#python3#Error / provider#python#Error

> It is for error debugging.

But then it should be define before the `finish`, right?

Shougo commented:
> But then it should be define before the finish, right?

Yes.

blueyed committed:
provider/python: define Prog and Error functions always

`provider#pythonX#Error` and `provider#pythonX#Prog` are currently meant
for debugging only (the error message is not being used), and should
therefore be defined always, especially in case of errors.

Ref: https://github.com/neovim/neovim/pull/2549#issuecomment-98607580

blueyed committed:
provider/python: Call: fix usage of 'finish' in function

Do not call it again in case of an exception in `remote#host#Require`
(ref: https://github.com/neovim/neovim/pull/2549#issuecomment-102674350).

blueyed commented:
There are several fixes in this PR by now.

Is it good like that, or should I split them into several PRs?

Pyrohh commented:
They all seem to apply to the Python provider, so it should be fine (although the PR should be renamed).

On May 5, 2015 5:38:30 AM EDT, Daniel Hahler <notifications@github.com> wrote:
>There are several fixes in this PR by now.
>
>Is it good like that, or should I split them into several PRs?
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/neovim/neovim/pull/2549#issuecomment-99011156

blueyed committed:
remote#host#RequirePythonHost: fix reference to log

I could not find a reference to NVIM_PYTHON_PYTHON_LOG anywhere, and
python-client looks for NVIM_PYTHON_LOG_FILE.

~/.nvimlog appears to be hardcoded and enabled by default.  This would
need to be adjusted when this changes.

Pyrohh commented:
@blueyed are all these commits individual fixes? If not, please squash some of them together.

blueyed commented:
@Pyrohh there is only one commit marked for fixup.
Should I squash it and then mark it RDY?

Pyrohh commented:
> Should I squash it and then mark it RDY?

I think that should be fine, but it looks like this has changed a lot since people looked at it, and I'm not knowledgeable in Vimscript. Could you take a look @Shougo / @splinterofchaos ?

Shougo commented on runtime/autoload/provider/pythonx.vim:
```diff
@@ -13,19 +13,21 @@ function! provider#pythonx#Detect(ver) abort
   let detect_versions = (a:ver == 2) ?
-        \   ['2.7', '2.6', '2', '']
-        \ : ['3.5', '3.4', '3.3', '3.2', '3', '']
+        \   ['2', '2.7', '2.6', '']
```
Why did you change the order?
I think you should not change the order.
https://github.com/neovim/neovim/pull/1334#issuecomment-94178512

blueyed replied:
The idea is to make this future proof and require less tests.
`python2` would typically point at the same version than `python2.7`.

@jonthn
Is it important in your case (from the comment at https://github.com/neovim/neovim/pull/1334#issuecomment-94178512) to test `2.7` and `2.6` before `2`?

jonthn replied:
@blueyed 
In my case as long as you test 2.6 and 2.7 before 'python' it would work (or in the worst case I can make it work without too much trouble and maintenance). Thanks for asking

blueyed replied:
Cool!
Then this seems to be fine, @Shougo.

Shougo replied:
OK.

fwalch replied:
> In my case as long as you test 2.6 and 2.7 before 'python' it would work (or in the worst case I can make it work without too much trouble and maintenance). Thanks for asking 

@jonthn It would work even if 'python' is tested before 2.7, as only 2.7 would have installed the `neovim` module, right?

jonthn replied:
@fwalch Yes, in my early issue (#1334) up to now it wouldn't have worked but now I have adopted a "saner" setup. And as long as fully specified version of python interpreter (e.g. python2.7) are checked it will work

blueyed commented:
@splinterofchaos
Can you please review this, and then change it to "RDY" / and/or merge it?

splinterofchaos commented:
The first three commits can be squashed as they operate on the same code. The rest are small commits; please think about whether any them can be squashed with the first three or grouped into one commit. Other than that, LGTM :+1:

blueyed committed:
provider/pythonx: return error from Detect

Ref: https://github.com/blueyed/neovim/commit/9b653ce0#commitcomment-11195449

blueyed commented:
I've rebased it on master, and addressed all comments.
I'd like to keep the commits separate for clarification, which would also allow for more detailed information when git-bisecting it etc.

Pyrohh commented:
> I'd like to keep the commits separate for clarification, which would also allow for more detailed information when git-bisecting it etc.

I think this is fine, but that's just me. Despite the small line count, this PR looks pretty dense with changes.

blueyed commented:
@splinterofchaos
I've noticed that using `echoerr` seems to be problematic, because it aborts any script, e.g. NeoBundle's sourcing and results in nasty errors:

    Vim(if):Channel was closed by the client
    Error detected while processing function UltiSnips#FileTypeChanged..provider#python3
    #Call:
    line    8:
    Failed to load python host. You can try to see what happened by starting Neovim with
     NVIM_PYTHON_LOG_FILE and opening the generated log file. Also, the host stderr will
     be available in Neovim log, so it may contain useful information. See also ~/.nviml
    og.
    Error detected while processing function neobundle#autoload#filetype..neobundle#conf
    ig#source_bundles..neobundle#config#source:
    line   58:
    E171: Missing :endif
    Error detected while processing function neobundle#autoload#filetype..neobundle#conf
    ig#source_bundles:
    line    2:
    E171: Missing :endif
    Error detected while processing function neobundle#autoload#filetype:
    line    4:
    E170: Missing :endfor
    Error detected while processing function neobundle#autoload#filetype..neobundle#conf
    ig#source_bundles..neobundle#config#source:
    line   56:
    E171: Missing :endif
    Error detected while processing function neobundle#autoload#filetype..neobundle#conf
    ig#source_bundles:
    line    2:
    E171: Missing :endif
    Error detected while processing function neobundle#autoload#filetype:
    line    4:
    E170: Missing :endfor

@Shougo
I think NeoBundle should maybe wrap the `dofiletype` in a `try`/`catch`.

For now, I've changed it to return in `Call` if there was some error before.
This should maybe get improved, so that `has("python3")` would fall back to returning `0` then?

The specific case currently is that I've manually broken my pyenv setup, but
I had a similar issue before, when there was some encoding error in the neovim Python module.

Shougo commented:
> I think NeoBundle should maybe wrap the doautocmd FileType in a try/catch 

I don't think so.  Why it is needed?

> The specific case currently is that I've manually broken my pyenv setup, but
 I had a similar issue before, when there was some encoding error in the neovim Python module.

Well, but to detect it is very slow.
It can detect in first call though.

equalsraf commented on runtime/autoload/provider/pythonx.vim:
```diff
@@ -39,31 +44,33 @@ function! s:check_version(prog, ver, skip) abort
         \ (a:ver == 2 ?
         \   '''import pkgutil; exit(pkgutil.get_loader("neovim") is None)''':
         \   '''import importlib; exit(importlib.find_loader("neovim") is None)''')
         \ )
```
Not necessarily for this PR, but since this python snippet is growing wouldn't it be better to save it as python_host_check.py in runtime and just `system(a:prog . ' $VIMRUNTIME/.../python_host_check.py')`? It could even print out the python-client version too.

blueyed commented:
Rebased/squashed again, and marking as "[RDY]" now.

Shougo commented:
When this fix is merged?
I think it is merged before #2735.

justinmk commented:
I'll wait for @fwalch or someone more familiar with this to merge it.

fwalch commented on runtime/autoload/provider/pythonx.vim:
```diff
@@ -12,26 +12,31 @@ function! provider#pythonx#Detect(ver) abort
 
   for prog in map(detect_versions, "'python' . v:val")
-    let [check, err] = s:check_interpreter(prog, a:ver, skip)
+    let [check, err, ver] = s:check_interpreter(prog, a:ver, skip)
```
Why add a return value (`ver`) here? It is not used.

blueyed replied:
You are right, it's not used currently, but could be used then for the main fix from https://github.com/neovim/neovim/pull/2735.

blueyed commented:
Rebased on master, amended the fixup-commit.

fwalch commented:
@blueyed Thanks. Can you squash your commits?

blueyed commented:
@fwalch 
See my previous comment (https://github.com/neovim/neovim/pull/2549#issuecomment-102115235).  I'd like to keep them separate.  They are separate fixes/improvements (as the PR evolved).

blueyed commented:
Is the build failure for `gcc` relevant? https://travis-ci.org/neovim/neovim/jobs/63944635

justinmk commented:
@blueyed It exceeded max time. Restarted. 

We're going to move away from valgrind in favor of clang SAN, to avoid these timeouts.

Pyrohh commented:
Besides what looks like another "standard" failure, everything else in Travis passed.

Pyrohh commented:
@fwalch (harmless bump, just in case you forgot about it)

ZyX-I commented:
> I've noticed that using echoerr seems to be problematic, because it aborts any script, e.g. NeoBundle's sourcing and results in nasty errors:

`:echoerr` either can emit unpredictable result (without immediate try block) or be an equivalent to `throw` (when it is immediately surrounded by try, so that caller knows that this will be an exception *always*). This is why I suggest to never use it. And if you do use it for some reason then you must use it in form `try | echoerr … | endtry` to make the results predictable.

Shougo commented:
> :echoerr either can emit unpredictable result (without immediate try block) or be an equivalent to throw (when it is immediately surrounded by try, so that caller knows that this will be an exception always). This is why I suggest to never use it. And if you do use it for some reason then you must use it in form try | echoerr … | endtry to make the results predictable.

Yes, I don't recommend to use `:echoerr`.  To use `:echomsg` is better.
If you want to print warning messages, you can change the highlight by `:echohl` command.

fwalch commented on runtime/autoload/provider/pythonx.vim:
```diff
@@ -39,31 +44,33 @@ function! s:check_version(prog, ver, skip) abort
-  let min_version = (a:ver == 2) ? '2.6' : '3.3'
-  if system(a:prog . get_version) >= min_version
+  let min_version = (a:ver[0] == 2) ? '2.6' : '3.3'
+  if a:ver >= min_version
```
This doesn't make sense. Of course `a:ver` will always be >= `min_version`, because `min_version` has been defined as `a:ver[0]`.

fwalch commented:
So `echoerr` should be replaced by `echomsg`? I guess the `finish` should be introduced back then as well?
I started to rebase #2735 on top of this to get rid of the merge conflicts. I got a bit confused, so if you got an notification email about a comment I posted a few minutes ago, please ignore it, I already deleted it. (Note to self: should probably start to edit comments to clarify instead of deleting them).

blueyed commented:
@fwalch 
> So `echoerr` should be replaced by `echomsg`? I guess the finish should be introduced back then as well?

Yes, that's what I would have done / understood.

It would make sense to use `echohl WarningMsg` here probably, too.

I assume you'll do this in your PR?  You can fixup the corresponding commit from me there.

Then I'd also say that we close this PR in favor of #2735?

fwalch commented:
@blueyed If that's okay with you, we can do it like this.. so I'd change d3294d7f5ae8442717df2d4cbc213bef0c163b6f back to use `echomsg` and add the `echohl` in there?

Shougo commented:
This issue is merged in #2735?
It is close-able?

blueyed commented:
@fwalch 
Yes. Maybe using a WarningMsg highlight.

Closing.
